### ISSUE TITLE
Poor macOS performance on discrete AMD graphics cards

### ISSUE THREAD

**Issue by StarWitch** [taisei-project/taisei#182]
I've talked with @Akaricchi about this privately but it's probably worth opening this up to others.

Currently, the default performance on macOS is quite poor using the default settings. Many sections of the game can't run beyond 20 fps, and some get to as low as 10 fps. Of particular note:

1. Stage 1 and 2 are slow essentially 100% of the time.
3. Oddly, Stage 5 works great at 55-60 fps, until a particular fairy enemy comes on-screen to fire "pink orb" danmaku with long tracer-tails in a circular pattern. Whatever is being used to render those particular danmaku bullets might be a good starting-point to investigate the slowdown. It's telling that the moment those danmaku go off screen and despawn, the game speeds up again.

However, when you disable all post-processing settings, the game springs back to life and is consistent at 60 fps, even in the sections mentioned above.

The easiest option would be to disable post-processing on macOS in the next release entirely, or to turn them off by-default, until it can be investigated further. Unfortunately, I have no experience debugging shaders or I'd offer to help out more.

Specs:
MacBook Pro 15" (2018)
Core i9, 32GB RAM, Radeon 560X
macOS Catalina 10.15.1

![pink-orb-danmaku](https://user-images.githubusercontent.com/34408664/70400148-45fc4a80-19f7-11ea-8f84-cc63330fd620.png)

**Comment by StarWitch** [comment]
Funny/sad update: on @Akaricchi's advice I went to see how far I could push the GPUs on my laptop, to see if it was some kind of "power-saving state" affecting performance. I used this fork of "gfxCardStatus" (https://github.com/steveschow/gfxCardStatus/releases/tag/v2.4.3i) to force my laptop to use a particular GPU.

It turns out that the integrated "Intel UHD Graphics 630" graphics chip runs the game at a full 60 fps with all post-processing enabled, but the Radeon 560X chugs on it for some reason. The integrated chip runs better than the discrete one.

Either way, I'm wondering if *something* can be done to make discrete graphics not run so poorly. There has to be something missing or incorrectly set to be causing these sorts of slowdowns. Hopefully those danmaku projectiles I mentioned above provide a clue.

**Comment by StarWitch** [comment]
Some more testing screenshots...

I noticed that some of the "effects" don't load on the Radeon 560X, while they seem to load just fine on Intel. This might also be a clue as to what shaders are causing this bug. 

This is the same build of the game with the same settings (post-processing set to full on all three settings), just switching between Radeon and Intel. 

![taisei-mac-slowdown-bug-water](https://user-images.githubusercontent.com/34408664/70641928-5e917e00-1c0c-11ea-8221-32cb7eabea9b.png)

**Comment by Akaricchi** [comment]
Hm. The Radeon screenshot is actually how it's supposed to look there.

**Comment by StarWitch** [comment]
OK, I think finally nailed it down. There was a lot of shader carnage and back-and-forth but I finally figured out the bottleneck for macOS performance with discrete AMD graphics (which is most Macbook Pros in the past 4 years).

Here are some links: 
1. https://bugs.chromium.org/p/chromium/issues/detail?id=735483
2. https://github.com/tsherif/WebGL/blob/f310f141ee1a634b85f154da2a979c4cb0dd2a09/sdk/tests/conformance/rendering/texture-switch-performance.html

The tl;dr is that `glUniform1i*` is simply not performant on macOS/AMD, for some reason. This is a long-standing issue going back to at least 2016, and instead of opening a bug with Apple over it, they've simply decided to tell people to use workarounds for the issue on this platform, which boils down to... 

```
"The workaround is to leave the sampler uniforms as is, and call bindTexture() / bindFramebuffer() to switch which texture is being sampled from and rendered to."
```

I ran their `glUniform1i*` test (linked above) against my own system, and can confirm that while Intel passed, AMD failed, giving similar numbers to Taisei. 

HOPEFULLY, this means we have enough info to fix this bug and make Taisei run well on macOS/AMD again.

**Comment by Akaricchi** [comment]
Wow. I'm speechless. This is just about the least expected issue ever. Ironically, Taisei's opengl renderer is actually optimized for the exact opposite! Because usually binding a texture is more expensive than simply changing a uniform.

Thanks for the investigation, I'll see what I can do.

**Comment by Akaricchi** [comment]
I've just pushed a potential fix in the `macos_amd_workaround` branch, and I need you to test a couple of things there:

1. Whether the issue detection works. With the native GL backend, you should see this warning in your log now: ```Enabled workaround `avoid sampler uniform updates` (buggy AMD driver on macOS)```. This currently should not automatically apply to ANGLE.
2. See how it affects performance, first in native OpenGL, and then in GLES3-over-OpenGL via ANGLE. You can force enable or disable the workaround with the `TAISEI_GL_WORKAROUND_AVOID_SAMPLER_UNIFORM_UPDATES` environment variable.

**Comment by StarWitch** [comment]
Yeah, it's a weird one. I have no idea what's happened but apparently AMD and Apple told the Chromium devs to just use their workaround. 

I ended up going into the shaders stubbing all of them into `fragColor = vec4(0.0, 0.0, 0.0, 0.0);` just to see what would happen. Ironically, the performance was identical even when literally all of the shaders were as basic as could be, even with nothing legible on the screen. Since it didn't, I ended up modifying a benchmarking script from sergeyext/FillRate to see where in the pipeline things would break, since I saw some mention of framebuffers being weird on recent AMD macOS drivers, but I had no luck there either. 

But with both the shaders and raw framebuffers being tested, I figured that there was possibly one or two GL calls in particular causing the issue, so I kept ripping stuff out until I got good performance.

I noticed that, oddly, performance would shoot up to 60 fps whenever I disabled background rendering. I decided to go into Stage 1's background rendering setup and started disabling certain drawing functions. Eventually, I found that `powersurge_effect()` was causing the most slowdowns, but ONLY on `draw_framebuffer_tex()` calls. Following the chain, it then went:

`stage1_bg`
`powersurge_effect()`
`draw_framebuffer_tex()/draw_framebuffer_attachment()`
`r_uniform_sampler()`
`shader_program.c (UNIFORM_SAMPLER type)`
`uset_int()`
`glUniform1iv()`

Commenting out that call made Taisei look really weird, but I was getting 60 fps. I then googled "glUniform1iv OpenGL AMD macOS" and immediately found that big thread on the Chromium mailing list with the exact same symptoms, and thankfully, a possible solution.

The reason why I had to do everything by hand like this was because OpenGL on macOS is a complete nightmare. macOS's Xcode OpenGL Profiler crashes on attachment to Taisei, and unfortunately renderdoc's macOS support has gotten stale and no longer functions (see: buldurk/renderdoc#1272 ). All other official OpenGL profiling apps simply don't work on macOS these days. I did, however, do something incredibly cursed, by running the WebGL/Emscripten build of Taisei in Chrome, and then running BabylonJS/Spector.js against it. It provided a somewhat useful dump, but it was still not enough to figure out what was happening at first glance.

----

As for the bugfix branch `macos_amd_workaround`, it works perfectly! I'm seeing the log message, as expected:

>W: glcommon_check_workaround: Enabled workaround `avoid sampler uniform updates` (buggy AMD driver on macOS; see https://github.com/taisei-project/taisei/issues/182)

Environment variable `TAISEI_GL_WORKAROUND_AVOID_SAMPLER_UNIFORM_UPDATES` also works as expected. Set to 0 it's as before, and letting it automatically detect AMD chips fixes the issue. For GLES3.0, which uses ANGLE, you'll need to modify `strstr(gl_vendor, "AMD")) && strstr(gl_renderer, "Radeon"))` to detect `gl_vendor, "Google"` in `detect_slow_sampler_update()`, since Google is the GL vendor for ANGLE.

----

I've learned more about OpenGL in the past two months than I ever thought I would. Thank you so much for entertaining my newbish questions, @Akaricchi!

**Comment by ksqsf** [comment]
Hi, I think this fix is worth a bugfix release. I hit the same bug, and I can confirm the problem is not present in master branch.

**Comment by Akaricchi** [comment]
@ksqsf There's an unfortunate reason why there haven't been any v1.3 minor releases since v1.3.1. Our previous writer has requested removal of her contributions for personal reasons (see #179). I have complied with the request and removed the old writing in master, and effectively promised that the next stable release will not have that content:
> Also please understand that we can not remove your contributions from the git history, or from the current and previous stable releases retroactively. The change has landed in the master development branch now, and will propagate to the next stable version (hopefully after we've replaced the stubs with actual story/dialogue).

We have found a new writer and completely replaced the old story and dialogue since then, but it's not a drop-in replacement. Some substantial changes have yet to be made to integrate it fully, which aren't appropriate for a bugfix release. So that leaves us with some pretty bad options:

- Release v1.3.2 with the old dialogue intact, violating the promise. I don't believe this is wrong from a legal perspective, but it's probably not a nice thing to do.
- Release v1.3.2 with the dialogue dummied out with no suitable replacement. This would just be a downgrade, at least for people not affected by major issues like the macOS problem.
- Do not release v1.3.2 at all, making people wait for v1.4 and/or apply the patches themselves.

So it's a tough call. So far I went with option 3. As a compromise, I can update the v1.3.x stable branch to include the fixes that would have gone into v1.3.2, but not tag an official release. You would still have to compile manually if you want the v1.3 experience without the bugs, but at least it'll be much easier to do than it is now.
